# Task 2.1: Data Pre-processing

In [1]:
# Importing necessary libraries
import seaborn as sns
from textblob import TextBlob
import sys
from wordcloud import WordCloud
import pandas as pd
import re
import numpy as np
import plotly
import plotly.express as px
import plotly.io as pio
import matplotlib.pyplot as plt
import tweepy,json
import numpy as np
import matplotlib.pyplot as plt

## Load the dataset

In [2]:
df = pd.read_csv('Downloads/Dataset-3/Bitcoin_tweets.csv')
df

/var/folders/gw/p4rv6sfn0fx4sk29djptzr_40000gp/T/ipykernel_12999/3421330669.py:1: DtypeWarning: Columns (4,5,6,7,12) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('Downloads/Dataset-3/Bitcoin_tweets.csv')


user_name    user_location  \
0                                   DeSota Wilson      Atlanta, GA   
1                                        CryptoND              NaN   
2                                       Tdlmatias  London, England   
3                            Crypto is the future              NaN   
4        Alex Kirchmaier 🇦🇹🇸🇪 #FactsSuperspreader           Europa   
...                                           ...              ...   
4132330                           All Things Defi              NaN   
4132331                               signal guy!    New York, USA   
4132332                        Blockchain World 🌐              NaN   
4132333                                   Tapez🇨🇦  Downtown Canada   
4132334                           All Things Defi              NaN   

                                          user_description  \
0        Biz Consultant, real estate, fintech, startups...   
1        😎 BITCOINLIVE is a Dutch platform aimed at inf...   
2        IM Academy : The best #forex, #SelfEducation, ...   
3        I will post a lot of buying signals for BTC tr...   
4        Co-founder @RENJERJerky | Forbes 30Under30 | I...   
...                                                    ...   
4132330                        #Defi lover, crypto hodler.   
4132331  Crypto investor and a learner !!!\nHelping fel...   
4132332  🚨Blockchain - crypto. $btc $alts and sometimes...   
4132333  Cryptocurrency Enthusiast and 1x Trader *Not f...   
4132334                        #Defi lover, crypto hodler.   

                user_created user_followers user_friends user_favourites  \
0        2009-04-26 20:05:09         8534.0         7605            4838   
1        2019-10-17 20:12:10         6769.0         1532           25483   
2        2014-11-10 10:50:37          128.0          332             924   
3        2019-09-28 16:48:12          625.0          129              14   
4        2016-02-03 13:15:55         1249.0         1472           10482   
...                      ...            ...          ...             ...   
4132330  2022-08-02 05:23:00          111.0          5.0             0.0   
4132331  2018-07-25 15:54:10         3894.0         13.0            26.0   
4132332  2011-12-04 04:43:27          387.0         89.0           946.0   
4132333  2019-12-21 21:53:45          203.0        223.0          1019.0   
4132334  2022-08-02 05:23:00          111.0          5.0             0.0   

        user_verified                 date  \
0               False  2021-02-10 23:59:04   
1               False  2021-02-10 23:58:48   
2               False  2021-02-10 23:54:48   
3               False  2021-02-10 23:54:33   
4               False  2021-02-10 23:54:06   
...               ...                  ...   
4132330         False  2022-09-02 19:18:51   
4132331         False  2022-09-02 19:18:26   
4132332         False  2022-09-02 19:18:16   
4132333         False  2022-09-02 19:18:04   
4132334         False  2022-09-02 19:17:48   

                                                      text  \
0        Blue Ridge Bank shares halted by NYSE after #b...   
1        😎 Today, that's this #Thursday, we will do a "...   
2        Guys evening, I have read this article about B...   
3        $BTC A big chance in a billion! Price: \487264...   
4        This network is secured by 9 508 nodes as of t...   
...                                                    ...   
4132330  updated price for $QTUM, $XLM, $SOL:\n\n#QTUM ...   
4132331  We did It 🎯\n\nBinance Futures\n#BTC/#USDT Tak...   
4132332  #btc 1h chart. Oversold and heading for 20.4 r...   
4132333  Possibly inv hns setup #btc https://t.co/Fazu9...   
4132334  updated price for $ADA, $AVAX, $TRX:\n\n#ADA $...   

                                                  hashtags  \
0                                              ['bitcoin']   
1                ['Thursday', 'Btc', 'wallet', 'security']   
2                                                      Na

## Filter Tweets with User Locations:

In [3]:
# Assuming 'user_location' is the column containing user locations
df_with_locations = df.dropna(subset=['user_location']).sample(n=500, random_state=42)

In [4]:
df_with_locations

user_name  \
4093421                                  Paul Schreibers   
624737                              Coin NEWS (Official)   
2082505                                   ℭ𝔯𝔶𝔭𝔱🎨𝔇𝔞🧛𝔞𝔪𝔭𝔛ℑ   
1419964                                Crypto Ambassador   
2656466                                      ChartingBot   
...                                                  ...   
2760250  د. خلود صالح المانع | Dr. Khulood Almani #MWC22   
3139061                                     Crypto Talks   
2250222                                     Asad Rasheed   
502944                                       ticker NEWS   
1614569                                      Favour Rose   

                         user_location  \
4093421               Perth, Australia   
624737           Tokyo,London, NewYork   
2082505                      Metaverse   
1419964                         Panama   
2656466                        Bitcoin   
...                                ...   
2760250        Kingdom of Saudi Arabia   
3139061  Our World , Blockchain Family   
2250222                            UAE   
502944      International News Channel   
1614569                            USA   

                                          user_description  \
4093421                            The Original Schreibers   
624737   Find the latest #cryptocurrency #news, updates...   
2082505  #DeepParody\nhttps://t.co/2yRSF6fPrM 🦄\nMind i...   
1419964  Decentralise everything! A fairer distribution...   
2656466  Daily crypto analysis for free. Developed by @...   
...                                                    ...   
2760250  #Top #Tech #World #Leader @getengati #CES2022 ...   
3139061  Follow us for Crypto Talks  \n#Crypto #Blockch...   
2250222  Financial Analyst, Follow me for the Real Time...   
502944   Streaming News. Now. 🌎 🖥 ticker NEWS covers th...   
1614569  NFT alerts, crypto alerts. We've got both. Che...   

                user_created user_followers user_friends user_favourites  \
4093421  2012-10-12 18:55:28             66           75               7   
624737   2021-05-11 12:46:24          103.0         36.0            11.0   
2082505  2008-04-05 16:37:39          536.0          185           95177   
1419964  2020-10-22 14:33:46          109.0        183.0          1173.0   
2656466  2022-01-21 15:23:53          100.0            1               1   
...                      ...            ...          ...             ...   
2760250  2021-06-12 23:48:32        31743.0       1977.0         76577.0   
3139061  2021-04-12 11:12:13          104.0        823.0           637.0   
2250222  2013-01-13 22:47:07         1499.0         19.0          2395.0   
502944   2019-08-10 09:22:24        10259.0        697.0          2978.0   
1614569  2021-10-20 15:17:22           22.0            2              41   

        user_verified                 date  \
4093421         False  2022-08-31 19:28:46   
624737          False  2021-07-30 18:04:03   
2082505         False  2022-01-12 15:57:14   
1419964         False  2021-10-22 20:48:02   
2656466         False  2022-03-21 20:45:07   
...               ...                  ...   
2760250         False  2022-04-16 09:50:27   
3139061         False  2022-04-30 15:44:16   
2250222         False  2022-01-20 08:44:55   
502944          False  2021-07-20 08:30:08   
1614569         False  2021-11-06 01:03:06   

                                                      text  \
4093421  What the #Bitcoin #Pizza is worth today! https...   
624737   The RC4 TestNet will become  $Neo N3 Formal Te...   
2082505  @ln_strike, again! People of 🇦🇷 get to experie...   
1419964  If you are into your #shiba #floki #kuma and a...   
2656466  📈Analysis March 21st📈\n\nThe current price of ...   
...                                                    ...   
2760250  @PDH_SciTechNews Thank you dear Philip @PDH_Sc...   
3139061  This dip is an ultimate buying opportunity\n\n...   
2250222  Recommendation\n#btcst | BUY\nEntry Price 1

## Text Preprocessing:

In [5]:
def preprocess_text(text):
    # Remove URLs
    text = re.sub(r'http\S+', '', text)
    # Remove special characters and symbols
    text = re.sub(r'[^a-zA-Z\s]', '', text)
    # Remove extra whitespaces
    text = re.sub(r'\s+', ' ', text).strip()
    return text

# Apply preprocessing to the 'text' column
df_with_locations['processed_text'] = df_with_locations['text'].apply(preprocess_text)

In [6]:
df_with_locations

user_name  \
4093421                                  Paul Schreibers   
624737                              Coin NEWS (Official)   
2082505                                   ℭ𝔯𝔶𝔭𝔱🎨𝔇𝔞🧛𝔞𝔪𝔭𝔛ℑ   
1419964                                Crypto Ambassador   
2656466                                      ChartingBot   
...                                                  ...   
2760250  د. خلود صالح المانع | Dr. Khulood Almani #MWC22   
3139061                                     Crypto Talks   
2250222                                     Asad Rasheed   
502944                                       ticker NEWS   
1614569                                      Favour Rose   

                         user_location  \
4093421               Perth, Australia   
624737           Tokyo,London, NewYork   
2082505                      Metaverse   
1419964                         Panama   
2656466                        Bitcoin   
...                                ...   
2760250        Kingdom of Saudi Arabia   
3139061  Our World , Blockchain Family   
2250222                            UAE   
502944      International News Channel   
1614569                            USA   

                                          user_description  \
4093421                            The Original Schreibers   
624737   Find the latest #cryptocurrency #news, updates...   
2082505  #DeepParody\nhttps://t.co/2yRSF6fPrM 🦄\nMind i...   
1419964  Decentralise everything! A fairer distribution...   
2656466  Daily crypto analysis for free. Developed by @...   
...                                                    ...   
2760250  #Top #Tech #World #Leader @getengati #CES2022 ...   
3139061  Follow us for Crypto Talks  \n#Crypto #Blockch...   
2250222  Financial Analyst, Follow me for the Real Time...   
502944   Streaming News. Now. 🌎 🖥 ticker NEWS covers th...   
1614569  NFT alerts, crypto alerts. We've got both. Che...   

                user_created user_followers user_friends user_favourites  \
4093421  2012-10-12 18:55:28             66           75               7   
624737   2021-05-11 12:46:24          103.0         36.0            11.0   
2082505  2008-04-05 16:37:39          536.0          185           95177   
1419964  2020-10-22 14:33:46          109.0        183.0          1173.0   
2656466  2022-01-21 15:23:53          100.0            1               1   
...                      ...            ...          ...             ...   
2760250  2021-06-12 23:48:32        31743.0       1977.0         76577.0   
3139061  2021-04-12 11:12:13          104.0        823.0           637.0   
2250222  2013-01-13 22:47:07         1499.0         19.0          2395.0   
502944   2019-08-10 09:22:24        10259.0        697.0          2978.0   
1614569  2021-10-20 15:17:22           22.0            2              41   

        user_verified                 date  \
4093421         False  2022-08-31 19:28:46   
624737          False  2021-07-30 18:04:03   
2082505         False  2022-01-12 15:57:14   
1419964         False  2021-10-22 20:48:02   
2656466         False  2022-03-21 20:45:07   
...               ...                  ...   
2760250         False  2022-04-16 09:50:27   
3139061         False  2022-04-30 15:44:16   
2250222         False  2022-01-20 08:44:55   
502944          False  2021-07-20 08:30:08   
1614569         False  2021-11-06 01:03:06   

                                                      text  \
4093421  What the #Bitcoin #Pizza is worth today! https...   
624737   The RC4 TestNet will become  $Neo N3 Formal Te...   
2082505  @ln_strike, again! People of 🇦🇷 get to experie...   
1419964  If you are into your #shiba #floki #kuma and a...   
2656466  📈Analysis March 21st📈\n\nThe current price of ...   
...                                                    ...   
2760250  @PDH_SciTechNews Thank you dear Philip @PDH_Sc...   
3139061  This dip is an ultimate buying opportunity\n\n...   
2250222  Recommendation\n#btcst | BUY\nEntry Price 1

# Task 2.2: Geocoding

In [ ]:
from geopy.geocoders import Nominatim
from tqdm import tqdm

# Initialize the geocoder
geolocator = Nominatim(user_agent="your_geocoding_app_name")

# Create functions to geocode locations
def geocode_location(location):
    try:
        location_info = geolocator.geocode(location)
        if location_info:
            return location_info.latitude, location_info.longitude
        else:
            return None, None
    except Exception as e:
        print(f"Error geocoding {location}: {e}")
        return None, None

In [ ]:
# Apply geocoding to the 'user_location' column
tqdm.pandas(desc="Geocoding")
df_with_locations[['latitude', 'longitude']] = df_with_locations['user_location'].progress_apply(geocode_location).apply(pd.Series)

In [7]:
# Print the DataFrame with added latitude and longitude columns
print(df_with_locations[['user_location', 'latitude', 'longitude']])


Python(13280) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Geocoding:   5%|█▋                             | 27/500 [00:24<19:43,  2.50s/it]

Error geocoding McDonald's: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=McDonald%27s&format=json&limit=1 (Caused by ReadTimeoutError("HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Read timed out. (read timeout=1)"))


Geocoding:   6%|█▋                             | 28/500 [00:30<28:01,  3.56s/it]

Error geocoding New Orleans, LA: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=New+Orleans%2C+LA&format=json&limit=1 (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x19a84a110>, 'Connection to nominatim.openstreetmap.org timed out. (connect timeout=1)'))


Geocoding:   6%|█▊                             | 29/500 [00:36<33:45,  4.30s/it]

Error geocoding Sivas, Türkiye: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Sivas%2C+T%C3%BCrkiye&format=json&limit=1 (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x19a849e70>, 'Connection to nominatim.openstreetmap.org timed out. (connect timeout=1)'))


Geocoding:   6%|█▊                             | 30/500 [00:42<37:45,  4.82s/it]

Error geocoding United States: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=United+States&format=json&limit=1 (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x19a849120>, 'Connection to nominatim.openstreetmap.org timed out. (connect timeout=1)'))


Geocoding:  10%|███▏                           | 51/500 [00:43<02:54,  2.57it/s]

Error geocoding Playa Del Carmen: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Playa+Del+Carmen&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a84afe0>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding Bitcoin City: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Bitcoin+City&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a84b5b0>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding The World Is Yours: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=The+World+Is+Yours&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a84bb80>: Fa

Geocoding:  18%|█████▌                         | 90/500 [00:43<00:41,  9.92it/s]

Error geocoding Puerto Rico, New York, Alaska : HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Puerto+Rico%2C+New+York%2C+Alaska+&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a849840>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding End of the rainbow.: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=End+of+the+rainbow.&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a8495a0>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding  🇺🇸🇯🇵 🇪🇺 🇮🇳 🇬🇧 🇨🇳: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=+%F0%9F%87%BA%F0%9F%87%B8%F0%9F%87%AF%F0%9F%87%B5+%F0%9F%87%AA%F0%9F%87%BA+%F0%9F%87%AE

Geocoding:  27%|████████▏                     | 136/500 [00:43<00:14, 25.88it/s]

Error geocoding Yonkers, NY: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Yonkers%2C+NY&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a84b100>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding The universe and beyond: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=The+universe+and+beyond&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a84ad70>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding Austin, TX: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Austin%2C+TX&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a84b640>: Fa

Geocoding:  36%|██████████▊                   | 181/500 [00:43<00:06, 50.86it/s]

Error geocoding Bedford, TX: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Bedford%2C+TX&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a848a30>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding Mars: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Mars&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a849780>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding Bengaluru South, India: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Bengaluru+South%2C+India&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a84a5c0>: Failed to establ

Geocoding:  44%|█████████████▎                | 221/500 [00:44<00:03, 81.23it/s]

Error geocoding France: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=France&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a8dcc40>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding Earth or Matrix ?: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Earth+or+Matrix+%3F&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a8b12d0>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding Web3: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Web3&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a8b0e20>: Failed to establish a new connection: 

Geocoding:  53%|███████████████▎             | 264/500 [00:44<00:01, 119.07it/s]

Error geocoding Japan: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Japan&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a8b3a00>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding San Francisco, CA: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=San+Francisco%2C+CA&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a84aec0>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding Ankara, Türkiye: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Ankara%2C+T%C3%BCrkiye&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a8b0130>: Failed to e

Geocoding:  62%|█████████████████▉           | 310/500 [00:44<00:01, 158.09it/s]

Error geocoding  Czech Republic: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=+Czech+Republic&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a84abf0>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding india: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=india&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a84a140>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding London, UK: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=London%2C+UK&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a848af0>: Failed to establish a new connec

Geocoding:  71%|████████████████████▌        | 354/500 [00:44<00:00, 178.81it/s]

Error geocoding RZA: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=RZA&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a84bd90>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding Knoxville, TN: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Knoxville%2C+TN&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a84a4a0>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding Global: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Global&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a84a1d0>: Failed to establish a new connection: [Errno 8] 

Geocoding:  81%|███████████████████████▎     | 403/500 [00:44<00:00, 206.55it/s]

Error geocoding Ninguna parte: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Ninguna+parte&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a84ad70>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding Ontario, Canada: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Ontario%2C+Canada&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a8dd960>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding Point Clear, England: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Point+Clear%2C+England&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a84b

Geocoding:  91%|██████████████████████████▍  | 456/500 [00:45<00:00, 231.22it/s]

Error geocoding Estonia: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Estonia&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a8b3610>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding Indianapolis ✈️ Houston: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Indianapolis+%E2%9C%88%EF%B8%8F+Houston&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a849330>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding Boston, MA: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Boston%2C+MA&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a84b01

Geocoding: 100%|██████████████████████████████| 500/500 [00:45<00:00, 11.03it/s]


Error geocoding Italy: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=Italy&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a84bca0>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding 21 Citadels: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=21+Citadels&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a849e70>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
Error geocoding MW-OA-SOL-3-39.4371N-76.6197W: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=MW-OA-SOL-3-39.4371N-76.6197W&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x19a8b1240>: Fail

In [8]:
df_with_locations

user_name  \
4093421                                  Paul Schreibers   
624737                              Coin NEWS (Official)   
2082505                                   ℭ𝔯𝔶𝔭𝔱🎨𝔇𝔞🧛𝔞𝔪𝔭𝔛ℑ   
1419964                                Crypto Ambassador   
2656466                                      ChartingBot   
...                                                  ...   
2760250  د. خلود صالح المانع | Dr. Khulood Almani #MWC22   
3139061                                     Crypto Talks   
2250222                                     Asad Rasheed   
502944                                       ticker NEWS   
1614569                                      Favour Rose   

                         user_location  \
4093421               Perth, Australia   
624737           Tokyo,London, NewYork   
2082505                      Metaverse   
1419964                         Panama   
2656466                        Bitcoin   
...                                ...   
2760250        Kingdom of Saudi Arabia   
3139061  Our World , Blockchain Family   
2250222                            UAE   
502944      International News Channel   
1614569                            USA   

                                          user_description  \
4093421                            The Original Schreibers   
624737   Find the latest #cryptocurrency #news, updates...   
2082505  #DeepParody\nhttps://t.co/2yRSF6fPrM 🦄\nMind i...   
1419964  Decentralise everything! A fairer distribution...   
2656466  Daily crypto analysis for free. Developed by @...   
...                                                    ...   
2760250  #Top #Tech #World #Leader @getengati #CES2022 ...   
3139061  Follow us for Crypto Talks  \n#Crypto #Blockch...   
2250222  Financial Analyst, Follow me for the Real Time...   
502944   Streaming News. Now. 🌎 🖥 ticker NEWS covers th...   
1614569  NFT alerts, crypto alerts. We've got both. Che...   

                user_created user_followers user_friends user_favourites  \
4093421  2012-10-12 18:55:28             66           75               7   
624737   2021-05-11 12:46:24          103.0         36.0            11.0   
2082505  2008-04-05 16:37:39          536.0          185           95177   
1419964  2020-10-22 14:33:46          109.0        183.0          1173.0   
2656466  2022-01-21 15:23:53          100.0            1               1   
...                      ...            ...          ...             ...   
2760250  2021-06-12 23:48:32        31743.0       1977.0         76577.0   
3139061  2021-04-12 11:12:13          104.0        823.0           637.0   
2250222  2013-01-13 22:47:07         1499.0         19.0          2395.0   
502944   2019-08-10 09:22:24        10259.0        697.0          2978.0   
1614569  2021-10-20 15:17:22           22.0            2              41   

        user_verified                 date  \
4093421         False  2022-08-31 19:28:46   
624737          False  2021-07-30 18:04:03   
2082505         False  2022-01-12 15:57:14   
1419964         False  2021-10-22 20:48:02   
2656466         False  2022-03-21 20:45:07   
...               ...                  ...   
2760250         False  2022-04-16 09:50:27   
3139061         False  2022-04-30 15:44:16   
2250222         False  2022-01-20 08:44:55   
502944          False  2021-07-20 08:30:08   
1614569         False  2021-11-06 01:03:06   

                                                      text  \
4093421  What the #Bitcoin #Pizza is worth today! https...   
624737   The RC4 TestNet will become  $Neo N3 Formal Te...   
2082505  @ln_strike, again! People of 🇦🇷 get to experie...   
1419964  If you are into your #shiba #floki #kuma and a...   
2656466  📈Analysis March 21st📈\n\nThe current price of ...   
...                                                    ...   
2760250  @PDH_SciTechNews Thank you dear Philip @PDH_Sc...   
3139061  This dip is an ultimate buying opportunity\n\n...   
2250222  Recommendation\n#btcst | BUY\nEntry Price 1

# Task 2.3 Polarity analysis

## Calculate Polarity:

In [9]:
# Assuming 'processed_text' is the column with preprocessed text
df_with_locations['polarity'] = df_with_locations['processed_text'].apply(lambda x: TextBlob(x).sentiment.polarity)

In [10]:
df_with_locations

user_name  \
4093421                                  Paul Schreibers   
624737                              Coin NEWS (Official)   
2082505                                   ℭ𝔯𝔶𝔭𝔱🎨𝔇𝔞🧛𝔞𝔪𝔭𝔛ℑ   
1419964                                Crypto Ambassador   
2656466                                      ChartingBot   
...                                                  ...   
2760250  د. خلود صالح المانع | Dr. Khulood Almani #MWC22   
3139061                                     Crypto Talks   
2250222                                     Asad Rasheed   
502944                                       ticker NEWS   
1614569                                      Favour Rose   

                         user_location  \
4093421               Perth, Australia   
624737           Tokyo,London, NewYork   
2082505                      Metaverse   
1419964                         Panama   
2656466                        Bitcoin   
...                                ...   
2760250        Kingdom of Saudi Arabia   
3139061  Our World , Blockchain Family   
2250222                            UAE   
502944      International News Channel   
1614569                            USA   

                                          user_description  \
4093421                            The Original Schreibers   
624737   Find the latest #cryptocurrency #news, updates...   
2082505  #DeepParody\nhttps://t.co/2yRSF6fPrM 🦄\nMind i...   
1419964  Decentralise everything! A fairer distribution...   
2656466  Daily crypto analysis for free. Developed by @...   
...                                                    ...   
2760250  #Top #Tech #World #Leader @getengati #CES2022 ...   
3139061  Follow us for Crypto Talks  \n#Crypto #Blockch...   
2250222  Financial Analyst, Follow me for the Real Time...   
502944   Streaming News. Now. 🌎 🖥 ticker NEWS covers th...   
1614569  NFT alerts, crypto alerts. We've got both. Che...   

                user_created user_followers user_friends user_favourites  \
4093421  2012-10-12 18:55:28             66           75               7   
624737   2021-05-11 12:46:24          103.0         36.0            11.0   
2082505  2008-04-05 16:37:39          536.0          185           95177   
1419964  2020-10-22 14:33:46          109.0        183.0          1173.0   
2656466  2022-01-21 15:23:53          100.0            1               1   
...                      ...            ...          ...             ...   
2760250  2021-06-12 23:48:32        31743.0       1977.0         76577.0   
3139061  2021-04-12 11:12:13          104.0        823.0           637.0   
2250222  2013-01-13 22:47:07         1499.0         19.0          2395.0   
502944   2019-08-10 09:22:24        10259.0        697.0          2978.0   
1614569  2021-10-20 15:17:22           22.0            2              41   

        user_verified                 date  \
4093421         False  2022-08-31 19:28:46   
624737          False  2021-07-30 18:04:03   
2082505         False  2022-01-12 15:57:14   
1419964         False  2021-10-22 20:48:02   
2656466         False  2022-03-21 20:45:07   
...               ...                  ...   
2760250         False  2022-04-16 09:50:27   
3139061         False  2022-04-30 15:44:16   
2250222         False  2022-01-20 08:44:55   
502944          False  2021-07-20 08:30:08   
1614569         False  2021-11-06 01:03:06   

                                                      text  \
4093421  What the #Bitcoin #Pizza is worth today! https...   
624737   The RC4 TestNet will become  $Neo N3 Formal Te...   
2082505  @ln_strike, again! People of 🇦🇷 get to experie...   
1419964  If you are into your #shiba #floki #kuma and a...   
2656466  📈Analysis March 21st📈\n\nThe current price of ...   
...                                                    ...   
2760250  @PDH_SciTechNews Thank you dear Philip @PDH_Sc...   
3139061  This dip is an ultimate buying opportunity\n\n...   
2250222  Recommendation\n#btcst | BUY\nEntry Price 1

## Average Polarity for Each Location:

In [11]:
average_polarity_by_location = df_with_locations.groupby(['latitude', 'longitude'])['polarity'].mean().reset_index()

In [12]:
average_polarity_by_location

latitude   longitude  polarity
0  -31.955893  115.860585  0.087500
1   -6.941114  107.666129  0.500000
2    8.559559  -81.130843  0.200000
3    9.064331    7.489297  1.000000
4   22.572646   88.363895  0.700000
5   23.764402   90.389015  0.707778
6   29.758938  -95.367697  0.200000
7   31.565682   74.314183  0.754000
8   34.053691 -118.242766  0.125000
9   37.016273   -7.935177 -0.288889
10  38.075005  126.455970  0.067500
11  40.712728  -74.006015  0.000000
12  40.789624  -73.959894  0.000000
13  41.947072  -87.655944  0.000000
14  46.732388 -117.000165  0.122222
15  48.428318 -123.364953  0.000000
16  51.225402    6.776314  0.200000
17  53.014822   18.610762  0.000000

## Geospatial Visualization:

In [13]:
import folium

# Create a map centered at a specific location
map_center = [average_polarity_by_location['latitude'].mean(), average_polarity_by_location['longitude'].mean()]
folium_map = folium.Map(location=map_center, zoom_start=2)

In [14]:
# Add markers for each location with average polarity
for index, row in average_polarity_by_location.iterrows():
    folium.Marker([row['latitude'], row['longitude']],
                  popup=f'Average Polarity: {row["polarity"]:.2f}').add_to(folium_map)

In [15]:
# Save the map as an HTML file or display it
folium_map.save('polarity_map.html')

- The resulting HTML file (polarity_map.html) will contain the geospatial visualization with markers indicating the average polarity for each location.

- Open the HTML file in a web browser to view the interactive map. 

# Task 2.4 Subjectivity Analysis

## Calculate Subjectivity:

In [16]:
# Assuming 'processed_text' is the column with preprocessed text
df_with_locations['subjectivity'] = df_with_locations['processed_text'].apply(lambda x: TextBlob(x).sentiment.subjectivity)

In [17]:
df_with_locations

user_name  \
4093421                                  Paul Schreibers   
624737                              Coin NEWS (Official)   
2082505                                   ℭ𝔯𝔶𝔭𝔱🎨𝔇𝔞🧛𝔞𝔪𝔭𝔛ℑ   
1419964                                Crypto Ambassador   
2656466                                      ChartingBot   
...                                                  ...   
2760250  د. خلود صالح المانع | Dr. Khulood Almani #MWC22   
3139061                                     Crypto Talks   
2250222                                     Asad Rasheed   
502944                                       ticker NEWS   
1614569                                      Favour Rose   

                         user_location  \
4093421               Perth, Australia   
624737           Tokyo,London, NewYork   
2082505                      Metaverse   
1419964                         Panama   
2656466                        Bitcoin   
...                                ...   
2760250        Kingdom of Saudi Arabia   
3139061  Our World , Blockchain Family   
2250222                            UAE   
502944      International News Channel   
1614569                            USA   

                                          user_description  \
4093421                            The Original Schreibers   
624737   Find the latest #cryptocurrency #news, updates...   
2082505  #DeepParody\nhttps://t.co/2yRSF6fPrM 🦄\nMind i...   
1419964  Decentralise everything! A fairer distribution...   
2656466  Daily crypto analysis for free. Developed by @...   
...                                                    ...   
2760250  #Top #Tech #World #Leader @getengati #CES2022 ...   
3139061  Follow us for Crypto Talks  \n#Crypto #Blockch...   
2250222  Financial Analyst, Follow me for the Real Time...   
502944   Streaming News. Now. 🌎 🖥 ticker NEWS covers th...   
1614569  NFT alerts, crypto alerts. We've got both. Che...   

                user_created user_followers user_friends user_favourites  \
4093421  2012-10-12 18:55:28             66           75               7   
624737   2021-05-11 12:46:24          103.0         36.0            11.0   
2082505  2008-04-05 16:37:39          536.0          185           95177   
1419964  2020-10-22 14:33:46          109.0        183.0          1173.0   
2656466  2022-01-21 15:23:53          100.0            1               1   
...                      ...            ...          ...             ...   
2760250  2021-06-12 23:48:32        31743.0       1977.0         76577.0   
3139061  2021-04-12 11:12:13          104.0        823.0           637.0   
2250222  2013-01-13 22:47:07         1499.0         19.0          2395.0   
502944   2019-08-10 09:22:24        10259.0        697.0          2978.0   
1614569  2021-10-20 15:17:22           22.0            2              41   

        user_verified                 date  \
4093421         False  2022-08-31 19:28:46   
624737          False  2021-07-30 18:04:03   
2082505         False  2022-01-12 15:57:14   
1419964         False  2021-10-22 20:48:02   
2656466         False  2022-03-21 20:45:07   
...               ...                  ...   
2760250         False  2022-04-16 09:50:27   
3139061         False  2022-04-30 15:44:16   
2250222         False  2022-01-20 08:44:55   
502944          False  2021-07-20 08:30:08   
1614569         False  2021-11-06 01:03:06   

                                                      text  \
4093421  What the #Bitcoin #Pizza is worth today! https...   
624737   The RC4 TestNet will become  $Neo N3 Formal Te...   
2082505  @ln_strike, again! People of 🇦🇷 get to experie...   
1419964  If you are into your #shiba #floki #kuma and a...   
2656466  📈Analysis March 21st📈\n\nThe current price of ...   
...                                                    ...   
2760250  @PDH_SciTechNews Thank you dear Philip @PDH_Sc...   
3139061  This dip is an ultimate buying opportunity\n\n...   
2250222  Recommendation\n#btcst | BUY\nEntry Price 1

## Average Subjectivity for Each Location:

In [18]:
average_subjectivity_by_location = df_with_locations.groupby(['latitude', 'longitude'])['subjectivity'].mean().reset_index()

In [19]:
average_subjectivity_by_location

latitude   longitude  subjectivity
0  -31.955893  115.860585      0.237500
1   -6.941114  107.666129      0.500000
2    8.559559  -81.130843      0.633333
3    9.064331    7.489297      1.000000
4   22.572646   88.363895      0.600000
5   23.764402   90.389015      0.837778
6   29.758938  -95.367697      0.200000
7   31.565682   74.314183      0.786000
8   34.053691 -118.242766      0.200000
9   37.016273   -7.935177      0.697222
10  38.075005  126.455970      0.332500
11  40.712728  -74.006015      0.000000
12  40.789624  -73.959894      0.800000
13  41.947072  -87.655944      0.500000
14  46.732388 -117.000165      0.394444
15  48.428318 -123.364953      0.000000
16  51.225402    6.776314      0.275000
17  53.014822   18.610762      0.575000

## Geospatial Visualization:

In [20]:
# Create a map centered at a specific location
map_center_subjectivity = [average_subjectivity_by_location['latitude'].mean(), average_subjectivity_by_location['longitude'].mean()]
folium_map_subjectivity = folium.Map(location=map_center_subjectivity, zoom_start=2)

In [21]:
# Add markers for each location with average subjectivity
for index, row in average_subjectivity_by_location.iterrows():
    folium.Marker([row['latitude'], row['longitude']],
                  popup=f'Average Subjectivity: {row["subjectivity"]:.2f}').add_to(folium_map_subjectivity)

In [22]:
# Save the map as an HTML file
folium_map_subjectivity.save('subjectivity_map.html')

# Task 2.5 Storify/Interpretation

As a policy adviser entrusted with informing the UK government regarding public opinion linked to cryptocurrencies throughout the world, the geographical analysis of sentiment including subjectivity across Twitter data gives significant insights. The research depends around a sample about 500 tweets including user locations, providing for a thorough grasp of worldwide opinions.

### Global Overview:

The geographical research demonstrates a broad variety of emotions towards cryptocurrencies throughout various areas of the globe. The sentiment scores, computed by polarity analysis, offer a quantifiable indication of the general sentiment in each place.

### Positive Views:

Certain localities display consistently good opinion towards cryptocurrencies. For instance, areas that have above average polarity ratings suggest strong favourable attitudes. Positive mood hotspots may signal good circumstances for bitcoin adoption and innovation. Notable favourable sentiment locations might possibly be investigated for collaborations, partnerships as well as trial initiatives relating to cryptocurrencies.

### Negative Views:

Conversely, localities that have lower average polarity ratings identify regions where antagonism towards bitcoin is more common. Understanding the factors behind the unfavourable attitude is vital for policymakers. Negative attitude could be related to problems such as regulatory challenges, security dangers, or public misunderstandings. Addressing these issues might be key for promoting good emotion and public acceptability.

### Mixed Sentiment:

Some locations display a mixed emotion, reflecting the balance among positive and negative opinions. Exploring the particular topics covered by tweets in these areas might uncover nuanced perspectives and areas of problem that require attention. Policymakers might have to develop a personalised strategy to handle both good and negative features.

### Considerations for Policy:

#### Location-Specific Impact: 
Given the worldwide nature of Twitter data, it's vital to evaluate the origin of tweets. Sentiments expressed in twitter starting outside of the UK may not directly impact UK government decisions. Considering the geographical context provides for a more sophisticated assessment of the data.

#### Importance of Clear Messaging: 
While assessing feelings, it's vital to examine if the messages are clear and coherent. Ambiguous or contradictory attitudes may signal a need for improved communication as well as public education regarding cryptocurrency-related concerns.

### Conclusion:

The geographical sentiment analysis is a significant tool for analysing the worldwide landscape of public sentiment on cryptocurrencies. Policymakers may exploit this knowledge to modify tactics, resolve issues and capitalize on favourable attitudes. The detailed insights gathered from this study assist to educated decision making along with proactive policy formulation in the dynamic realm of cryptocurrencies.